client/client-lib/src/modules.rs

```diff
@@ -0,0 +1,481 @@
+}
+
+#[derive(Encodable, Decodable)]
+pub struct Signature;
+
+#[derive(Debug, Error)]
+pub enum TransactionError {
+    #[error("your inputs are too low")]
+    NotEnoughInputs,
+}
+
+impl UnsignedTransaction {
+    /// Generate change outputs using `module`
+    fn add_change(
+        &mut self,
+        _module: &FedimintModule,
+        _rng: &mut DeterministicRandomnessTracker,
+    ) -> Result<Vec<SpendableOutput>, TransactionError> {
+        todo!()
+    }
+
+    /// Sign into [`Transaction`] and return new [`SpendableOutput`]s
+    fn sign(
+        &self,
+        _rng: &mut DeterministicRandomnessTracker,
+    ) -> (Transaction, Vec<SpendableOutput>) {
+        // check if inputs == outputs
+        // validate input sigs?
+        // calculate txid
+        // convert all the `SpendableOutput`s to `Inputs`
+        // use keys to sign tx
+        todo!()
+    }
+}
+
+/// Transaction that was already signed
+#[derive(Encodable)]
+pub struct Transaction {
+    inputs: Vec<Input>,
+    outputs: Vec<Output>,
+    signature: Signature,
+}
+
+impl<M> ModuleDecodable<M> for Transaction
+where
+    Input: ModuleDecodable<M>,
+    Output: ModuleDecodable<M>,
+{
+    fn consensus_decode<R: std::io::Read>(
+        r: &mut R,
+        modules: &BTreeMap<ModuleKey, M>,
+    ) -> Result<Self, DecodeError> {
+        Ok(Self {
+            inputs: ModuleDecodable::consensus_decode(r, modules)?,
+            outputs: ModuleDecodable::consensus_decode(r, modules)?,
+            signature: Decodable::consensus_decode(r)?,
+        })
+    }
+}
+
+pub trait IFedimintModule {
+    fn module_key(&self) -> ModuleKey;
+
+    /// Generate a transaction [`Output`] and
+    fn generate_output(
+        &self,
+        amount: Amount,
+        deterministic_rand: &bitcoin::hashes::sha256::Hash,
+    ) -> (Output, SpendableOutput);
+
+    /// Decode `SpendableOutput` compatible with this module, after the module key prefix was already decoded
+    fn decode_spendable_output(&self, d: &mut dyn io::Read)
+        -> Result<SpendableOutput, DecodeError>;
+
+    fn decode_output(&self, d: &mut dyn io::Read) -> Result<Output, DecodeError>;
+    fn decode_input(&self, d: &mut dyn io::Read) -> Result<Input, DecodeError>;
+}
+
+#[derive(Clone)]
+#[autoimpl(Deref using self.0)]
+pub struct FedimintModule(Arc<dyn IFedimintModule + Send + Sync + 'static>);
+
+// autoimpl below expands to something like:
+// impl Deref for MintModule {
+//     type Target = dyn IMintModule + Send + Sync + 'static;
+
+//     fn deref(&self) -> &Self::Target {
+//         &*self.0
+//     }
+// }
+
+pub struct MintModule {
+    #[allow(unused)]
+    config: MintModuleConfig,
+}
+
+pub struct MintModuleConfig {
+    #[allow(unused)]
+    some_number: u32,
+}
+
+impl MintModule {
+    #[allow(clippy::new_ret_no_self)]
+    pub fn from_config(config: MintModuleConfig) -> FedimintModule {
+        FedimintModule(Arc::new(Self { config }))
+    }
+}
+
+impl IFedimintModule for MintModule {
```

Comment:
Is there anyway to avoid the encode/decode boilerplate?  Maybe a custom macro?

Reply:
Boilerplate can always be reduced via basic or proc macros. But there's no escaping the fact that it's additional work and complexity.

Hopefully there is going to be that many types affected. Basically: `Input`, `Output`, `PendingOutput`, `SpendableOutput`, `FedimintModule` implemented for every module.

Also these macros would probably be slightly altered copies of exting `fedimint-derive` macros.